### PR TITLE
33 firebase error handler

### LINKED ISSUE #33: create firebase error handler
whenever get a firebase request error, show the user a snackbar or dialog

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes #33 

Hi nick, went out of scope of ticket by implementing firebase error handling in signinscreen. 
I did that so i could write tests for it.

i also had to touch the main,dart file, setting a navService in main(), and feeding the navigaorKey to myapp.
cant remember why we made the authservice in the main() function and not my app but i dont doubt theres a good reason haha.
and i updated the classes in your test files, if the ci tests pass, then im a lucky man. 

**Not sure if will work on ios**
- [x] can you please test on ios for me, im not sure if the error message codes will be the same

### CODE REVIEW

**Comment by nickmeinhold** [comment]
Hey mate!  Just having a look through now.  

> cant remember why we made the authservice in the main() function and not my app 

It's so we can inject mock services into the MyApp widget - either in a widget test where the widget under test is MyApp (so effectively the whole app) or in integration tests. 

> can you please test on ios for me, im not sure if the error message codes will be the same 

What specifically do you want me to test? What error message code are you expecting?

File: lib/screens/sign_in.dart [hunk 1]
```diff
@@ -105,14 +107,25 @@ class _SignInScreenState extends State<SignInScreen> {
...
                               });
                             } else {
                               _autoValidateForm = true;
                             }
+
+                            final authResult = await context
+                                .read<AuthService>()
+                                .signInWithEmail(_email, _password);
+
+                            // stop loading animation
```
Comment (nickmeinhold) [inline lib/screens/sign_in.dart:123]: Do you remember when I was saying there was something about this that bothered me? I realised what it was I think - I played around with different ways of doing this login process and ended up having a piece of state, called (eg.) authState that the UI observes and the services can update.  It becomes more useful when there are several stages, like in the AuthProvider flows.  Anyway no need to change now imo, just wanted to flag it for discussion at some point.

File: lib/services/auth/firebase_auth_service.dart [hunk 2]
```diff
@@ -21,9 +24,44 @@ class FirebaseAuthService implements AuthService {
...
-    final authResult = await _firebaseAuth.signInWithEmailAndPassword(
-        email: email, password: password);
-    return authResult.user.toData();
+    try {
+      final result = await _firebaseAuth.signInWithEmailAndPassword(
+          email: email, password: password);
+      return result.user.toData();
+    } on PlatformException catch (error) {
+      String customErrorMessage;
+      switch (error.code) {
```
Comment (nickmeinhold) [inline lib/services/auth/firebase_auth_service.dart:33]: Oh right, this is the error codes you mentioned.  I ran the app on an iOS simulator and created an account - there was no exception or error message displayed, it just stayed at the 'thinking' animation.  

I tried checking the db but I don't have access (which is fine with me just giving you the rundown of what I did) 

Screen Shot 2020-08-12 at 1 11 42 am

Comment (nickmeinhold) [reply lib/services/auth/firebase_auth_service.dart:33]: Sorry, I see this is just for the sign in, not the create account.  I tried signing in and got the expected alert about no such user found.

File: lib/services/auth/navigation_service.dart [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  }
+
+  void displayError(dynamic error) {
+    final message = error.toString();
+
+    // filter out messages user doesnt want to see here
+
+    // show dialog
+    final newContext = navigatorKey.currentState.overlay.context;
+    showDialog<dynamic>(
```
Comment (nickmeinhold) [inline lib/services/auth/navigation_service.dart:18]: One thing that I forgot about when suggesting this strategy is that calling showDialog either during a build or not during a build... I can't remember the details but there can be weird subtle issues.  I've got an issue tracking working on it: https://github.com/nickmeinhold/crowdleague/issues/110  

I think it just needs a little beefing up to deal with all scenarios, again I reckon don't worry about it now and I'll let you know what I come up with when I get that issue done.

**Review by nickmeinhold — APPROVED** [review]
Looks great man, awesome work!!!

**Comment by nickmeinhold** [comment]
@tamari-gray just to follow up on the exceptions question, the exceptions are (as far as I understand) consistent across platforms since https://github.com/flutter/plugins/pull/775 and apparently are being improved to use types: https://github.com/FirebaseExtended/flutterfire/issues/2653

File: lib/screens/sign_in.dart (hunk 1 above)
Comment (tamari-gray) [reply lib/screens/sign_in.dart:123]: really keen to talk about it nick! I was working on a solution that would work like that and kept running into bugs all week, still couldnt really wrap my head around it so i went with the solution we worked on to get some momentum (productivity wise). Shouldve asked for some help earlier but still learned abit :) 
Lets chat tommorow! Ill be at the meetup, or just anytime that suits

File: lib/services/auth/firebase_auth_service.dart (hunk 2 above)
Comment (tamari-gray) [reply lib/services/auth/firebase_auth_service.dart:33]: thank you, sorry ill be more specific next time :)

**Comment by tamari-gray** [comment]
@nickmeinhold thanks so much for the detailed comments/ explanations & feedback!

File: lib/screens/sign_in.dart (hunk 1 above)
Comment (nickmeinhold) [reply lib/screens/sign_in.dart:123]: Ok brilliant!  Let's get stuck into it tomorrow :-)

**Pull request merged**